architecture/design/multi-region-2DC-deployment.md

```diff
@@ -216,13 +216,61 @@ Upon receiving a `ReplicateMsg` that is a part of a multi-row transaction, the *
+
+For the time-based policy, the Log object also uses a private variable `wal_retention_secs_` which gets modified through an `ALTER TABLE` request. When deciding if a log file is old enough to be deleted, we use `max(log_min_seconds_to_retain, wal_retention_secs_)`.
+
+To avoid deleting log segments that contain unreplicated entries, we will add a new policy based on the minimum op id that we want to retain. This new policy will be disabled by default and will be controlled by the gflag `enable_log_retention_by_op_idx`.
+
+Because of the possibility of very long network partitions or unused cdc streams that were not deleted, we will have an option in which the op id policy is ignored if the tserver is running out of disk space. If the gflag `log_stop_retaining_min_disk_mb` is set, we will stop retaining records with unreplicated entries once disk space is less than the specified value. This gflag will not affect log segments that are being retained because its entries haven’t been flushed to rocksDB.
+
+We will also provide a new gflag `log_max_seconds_to_retain` to specify the maximum retention time. If a log file is older than this gflag, we will delete it even if one if its records hasn’t been replicated, but only if deleting such file doesn’t violate any of the other policies specified by `log_min_seconds_to_retain` and `log_min_segments_to_retain`. If log_max_seconds_to_retain is set to a value less than or equal to `max(log_min_seconds_to_retain, wal_retention_secs_)`, it will be ignored.
+
+## Last replicated op id
+Our 2DC implementation can only replicate entries that are in the log files (it’s not possible to replicate entries that are only in rocksDB), so it’s important that we don’t delete log files that contain unreplicated entries.
+
+When a consumer polls for changes, it sends to the producer the latest replicated op id, and this information is saved on the producer's `cdc_state` table. The producer will take this information and call `Log::SetMinimumOpIdToRetain()`. This will in turn set the private variable `minimum_op_idx_to_retain_` which will be used by the Log garbage collector only when enable_log_retention_by_op_idx is set to true.
+
+`Log::SetMinimumOpIdToRetain()` will be called every time a cdc stream’s last replicated op id is updated (in `CDCServiceImpl::UpdateCheckpoint`) and every time the cdc service starts.
+
+## Setting up the op id retention policy
+For this policy to be active, all the tservers in the producing universe will have to be started with `--enable_log_retention_by_op_idx=true`. Initially, if no cdc stream is setup for a specific table, the policy will remain disabled for all the tablets in the table. Once the first cdc stream is setup, `Log::SetMinimumOpIdToRetain()` will be called with a value of 0. This will effectively disable garbage collection of log files until the consumer updates the latest replicated op id.
```

Comment:
```suggestion
For this policy to be active, all the tservers in the producer universe will have to be started with `--enable_log_retention_by_op_idx=true`. Initially, if no cdc stream is setup for a specific table, the policy will remain disabled for all the tablets in the table. Once the first cdc stream is setup, `Log::SetMinimumOpIdToRetain()` will be called with a value of 0. This will effectively disable garbage collection of log files until the consumer updates the latest replicated op id.
```